src/cryptography/x509/name.py

```diff
@@ -4,14 +4,45 @@
 
 from __future__ import absolute_import, division, print_function
 
+from enum import Enum
+
 import six
 
 from cryptography import utils
 from cryptography.x509.oid import NameOID, ObjectIdentifier
 
 
+class _ASN1Type(Enum):
```

Comment:
This set of types was chosen because it's the set that OpenSSL will *not* error on when calling `ASN1_STRING_to_UTF8`. If we don't represent the complete set here then it's possible to raise errors where we previously succeeded.

Reply:
See: https://github.com/openssl/openssl/blob/c0dba2cca4d2bf3526d90a2050bdb17148ce803f/crypto/asn1/a_strex.c#L281-L290

Reply:
Why do we need so many types? I thought we only really wanted `UTF8String` and `PrintableString?`

Reply:
When decoding a cert we look up the type from the dict that maps back to enum. If we don't represent the complete set of types that OpenSSL handles with `ASN1_STRING_to_UTF8` then we'll have a regression where certs that previously parsed will now fail with a `KeyError` when encountering an unknown ASN1 string type. This list is the complete set OpenSSL supports.

Reply:
Hmm, ok on further thought, the set of types we really need here is the set of types you might get out of `X509_NAME_ENTRY_get_data`, is that the same?

Reply:
`X509_NAME_ENTRY_get_data`'s implementation just dereferences and returns the ASN1_STRING *, so it can be any value for `type` in the asn1_string_st struct (which is just an int). So I think that could technically be any ASN1 type 0-30?

We could reorder the code so that the type lookup occurs after we do the `ASN1_STRING_to_UTF8` call, which would limit it to the set we've got defined here.